UIAComWrapper Input Devices Refactoring

dpisanu opened the pull request:
Cleaning up the Keyboard components

BareMetalKeyboard

This contains the DLLImports from user32
Basic Keyboard interaction and input validation
IBaseKeyboard / BaseKeyboard

Moved alot of the Keyboard Functions and Properties from Keyboard to
BaseKEyboard
BaseKeyboard handles anything not relevant to an IActionListener
IKeyboard / Keyboard

Handles anything related to an IActionListener
AttachedKeyboard

Implements IBaseKeyboard
Handles focusing on an UI Item Container before doing any keyboard
interaction

--- --- ---

Cleaning up the Mouse Classes and Interfaces
This 'should' not break current implementation.
It's just separating some of the code into manageable chunks.

BareMetalMouse

Added new Class called BareMetalMouse
Contains the DLLImport from user32
Contains the basic Button Interactions
IBaseMouse / BaseMouse

New Interface called IBaseMouse
New Abstract Class called BaseMouse
Moved most of the Functions from Mouse to BaseMouse
Interface now reflects all the public Functions and Properties from the
Class
BaseMouse contains all interactions that do not directly require an
ActionPerformed on the IActionListener
IMouse / Mouse

Cleaned up Mouse -> Move the Bare Mouse work to BareMetalMouse
Cleaned up Mouse -> Moved general Mouse Work to BaseMouse
Mouse now only contains the code for Mouse Interaction sporting an
IActionListener
Interface reflects the public Functions and properties
AttachedMouse

Inherits from IBaseMouse now
Offering more Function Calls now
On Operations where Mouse offers a version with IActionListener, that
Function is called
On Operations where Mouse does not offer an IActionListener call,, the
ActionPerformed Function from Mouse is called

Roemer commented:
Just for information (I will review the whole thing later): I think we can de-release obsolete methods in the uiacom branch and focus on getting a clean codebase.

Roemer commented on src/TestStack.White.UITests/WhiteUITestBase.cs:
```diff
@@ -81,7 +88,7 @@ private IDisposable SetMainWindow(WindowsFramework framework)
             try
             {
-                Keyboard = Keyboard.Instance;
+                Keyboard = new Keyboard();
```
Do we need a new Keyboard here?

Roemer commented on src/TestStack.White/InputDevices/AttachedMouse.cs:
```diff
@@ -8,62 +9,250 @@ namespace TestStack.White.InputDevices
+        /// <summary>
+        /// Implements <see cref="IBaseMouse.Click()"/>
+        /// </summary
+        [Obsolete("Use LeftClick instead")]
```
As already mentioned, I think we can delete obsolete methods.

dpisanu commented:
Updated PR. 
Removed the Obsoletes. I had planned to get them with the next cleanup PR i have in my pocket but might as well take them out with this refactoring. 

The other comment has been taken care off as well.

Roemer commented:
Do we really need that many classes? Do we  for example need IBaseKeyboard, isn't BaseKeyboard enough?
From a quick look I think BareMetalKeyboard, BaseKeyboard, Keyboard and AttachedKeyboard is enough, same for mouse.

dpisanu commented:
As I explained before to Jake already.
Basing on the design that was there already, 
AttachedMouse and AttachedKeyboard belong to a specific UIItemContainer.
They work exactly for this Container, where as Mouse and Keyboard are more general and not Item Specific.

If you remove the Attached versions, you will need to take care of the IActionListener for most of the Mouse Calls for an UIItem and on the Keyboard side the Focus step.
So, in a way, having the Attached Versions wrap the general implementations and let them handle their UIItem is not THAT bad. 
Saves having to think about setting the correct IActionListener and always remembering to Focus before Keyboard actions.

The reason why there is a split between BaseMouse and Mouse, is that the concrete Mouse Implementations were using the same Interface.
By putting all the functions from the Class into the Interface we were left with too many Functions that were not relevant in the AttachedMouse.
So I boiled it down to common denominator and from that create the Base Version.

JakeGinnivan commented on src/TestStack.White/InputDevices/AttachedMouse.cs:
```diff
@@ -8,62 +9,211 @@ namespace TestStack.White.InputDevices
+        
+        /// <summary>
+        /// Implements <see cref="IBaseMouse.LeftClick()"/>
+        /// </summary
```
Missing close

JakeGinnivan commented on src/TestStack.White/InputDevices/Mouse.cs:
```diff
@@ -1,145 +1,24 @@
-        [DllImport("user32.dll")]
-        static extern int GetSystemMetrics(SystemMetric smIndex);
-
-        public static Mouse Instance = new Mouse();
```
Removing this is way too big a breaking change, look at the impact it had *within* this pull request.

This is a great way to drive users away from a library

dpisanu replied:
I thought this was the idea. Allow breaking changes in the UIAComWrapper branch.
Even breaking changes in the API.

From a personal opinion, purely speaking as a customer of White, a stale project is more of a reason to ditch and leave then API changes.
API changes are a sign of change and leads to a maturing product.

dpisanu replied:
On a side not, I see Fields as Class-Level.
The way the static is being used is more on the side of a Singleton.
That raises the question. Do you want a Singleton or do you want Mouse to be also instanciated?
Serve one or the other.

I would be ok as to making Mouse and Keyboard a singleton pure and through.

JakeGinnivan commented:
Even after explaining I still don't get it.

`AttachedKeyboard`/`AttachedMouse` are decorators, they have no common base class or common functionality with the original? They currently implement the decorator pattern.

Also, UIItem should always return a `AttachedKeyboard`, never an IKeyboard.

dpisanu commented:
From an IOC perspective, which is valid scenario (that I am currently using in my wrapping of White), I would rather have the Interfaces exposed and never the concrete implementations.

Then the decorator should also have it's own interface.
This will lead to code duplication since AttachedMouse and Mouse will an equal set of functions defined in two different Interfaces.
If you are ok with such a case of code duplication, fine. I can do the the change.

dpisanu commented:
So I just cut down the classes.
Removed the BaseMouse and BaseKeyboard.

If AttachedMouse and AttachedKeyboard are just 'decorators' and should have no common Interface with Mouse and Keyboard, they can be decoupled.
I added Interfaces for AttachedMouse and AttachedKeyboard though.

Now the last question that is open is the use of the static Instance field.
```javascript
public static readonly Keyboard Instance = new Keyboard();

public static Mouse Instance = new Mouse();
```

Do you still want this in?
Would it make more sense to change Keyboard and Mouse to be Singletons then?

Roemer commented:
I like the newest changeset. As for the Instance/Singleton: In my opinion this should be avoided when ever possible. Either make it completely static or create some sort of "provider" which provides an unique instance of this class.

dpisanu commented:
So maybe let's get this in first and then address the "static" access topic in a second Pull Request that can be planned out.
I see a difference of opinion here on maintainer level that should be cleared out.

JakeGinnivan commented:
> I see a difference of opinion here on maintainer level that should be cleared out.

I don't disagree with some of the static vs Instance property suggestions. I just don't see the value in making a breaking change. It looks like just making changes for the sake of it, without weighing anything up. My preference is for static, OR simply pass a configuration object into the `Application` when you new it up. That said, I don't think the change brings much value so I wouldn't personally make it.

Quick question, are both of you happy to be available to handle issues people raise when the breaking changes go out? What about if the refactorings slightly change behaviors?

Currently the release notes for the next version will be (if these changes were released on master). We broke some stuff because its good practice, we didn't actually fix anything :trollface:

Roemer commented:
The codebase of White is old, you can see that almost everywhere when browsing it. I think this is one of the reasons that there's not much activity in White. Personally I'd prefer to have an active development in White than keep the current users happy (they don't need to upgrade anyway). And to get this, we first need to get the codebase clean. So in my opinion, the next release should be the UIAComWrapper with a clean code base. Then we have a good foundation to get White back to life and keep it's development running. Everything else is just "keeping it alive as long as possible".
I don't say API breaking changes are great. But I think we now have a chance to clean it up and have a good foundation for the future.

dpisanu commented:
I think once the UIAComWrapper Branch is released as the next version, we are able to say "We fixed things AND made the code base better because it's good practice)

Roemer commented:
From my point of view, this changeset can be merged.

dpisanu commented:
Yay!

JakeGinnivan commented:
Ok cool. Also happy to add you as a maintainer @dpisanu. Appreciate you getting involved :) I do have some suggestions for changes though ;P will comment inline

JakeGinnivan commented on src/TestStack.White/InputDevices/AttachedMouse.cs:
```diff
@@ -8,62 +8,211 @@ namespace TestStack.White.InputDevices
+        
+        /// <summary>
+        /// Implements <see cref="IAttachedMouse.LeftClick()"/>
+        /// </summary
```
Most of the summary tags are missing a `>`

dpisanu replied:
These elusive closing tags ... done

JakeGinnivan commented on src/TestStack.White/InputDevices/BareMetalKeyboard.cs:
```diff
@@ -0,0 +1,84 @@
+    /// <summary>
+    /// Represents Keyboard attachment to the machine.
+    /// </summary>
+    public static class BareMetalKeyboard
```
Everything in this class is internal, make this internal and rename to `KeyboardWin32`?

dpisanu replied:
Sounds good.

JakeGinnivan commented on src/TestStack.White/InputDevices/BareMetalMouse.cs:
```diff
@@ -0,0 +1,119 @@
+
+namespace TestStack.White.InputDevices
+{
+    public static class BareMetalMouse
```
Ditto, internal class and `MouseWin32`?

dpisanu replied:
Sounds good.

JakeGinnivan commented on src/TestStack.White/InputDevices/Keyboard.cs:
```diff
@@ -1,193 +1,230 @@
-
-        private readonly List<int> keysHeld = new List<int>();
+        [Obsolete("Obsolete. Create an instance where it is needed.")]
+        public static readonly IKeyboard Instance = new Keyboard();
```
I don't mind having a default instance available. It saves on allocations and the majority of the time you want to access this as static?

Or do you think the majority access they mouse/keyboard via UIItems?

dpisanu replied:
I am very torn on this.
What I do not want is a mixture of the two possibilities.

On one side, Mouse and Keyboard are things that can be accessed via a static instance.
Since Mouse and Keyboard are the same system wide.

On the other hand, it could make sense to allow Users to create their own Mouse and Keyboard implementations that adhere to the interface but with a custom workflow inside.

If you look at unit testing, persistent states are really bad when it comes to testing since tests should IDEALLY be independent of each other.

I would like to postpone this to a later point to let some more thoughts go into it.

JakeGinnivan replied:
:+1:

JakeGinnivan commented on src/TestStack.White/InputDevices/IKeyboard.cs:
```diff
@@ -1,13 +1,81 @@
+        /// </summary>
+        /// <param name="keysToType">String to send</param>
+        /// <param name="actionListener"><see cref="IActionListener"/> to receive the string</param>
+        void Send(string keysToType, IActionListener actionListener);
```
These seem like an internal concept really? We could instead remove all `IActionListener` overloads from the interface, then make `AttachedKeyboard`'s constructor take a `IActionListener`.

Then in `UIItem` can use `new ActionListenerKeyboardDecorator(new Keyboard(), actionListener)`. That way the interfaces are not bloated, and `IAttachedKeyboard` can just inherit from `IKeyboard`

dpisanu replied:
As said up ahead. It could make sense to allow customers to create their own Keyboard and/or Mouse implementations that adhere to the Interface and use the IActionListener in their own manner.

JakeGinnivan replied:
As far as I know, the main reason for the IActionListener is for recording what is happening.

If the user passes a custom IActionListener to the UIItem then it will be used with the keyboard decorator. OR we provide an overload to Keyboard and it calls if it's required. Both of these remove the duplication of members on `IAttachedKeyboard` and `IKeyboard` and also remove all the methods taking an IActionListener?

JakeGinnivan commented:
Let me know your thoughts on my suggestions.

dpisanu commented:
@JakeGinnivan I put in some of your review comments.
I would be much obliged to be a maintainer.
I think we could be getting some momentum here to get the ball rolling again in the right direction.
Albeit that I haven't touched UIA in it's newest Version, I am more then happy to assist in any way I can.

JakeGinnivan commented:
Looking good. Still wondering if we can remove the IActionListener overloads. It would clean things up, and TBH I can't see anyone really using them directly

JakeGinnivan commented:
Or just remove them from the interface, and create another interface then implement that interface explicitly in the class. Then if anyone needs them externally they just cast to the IKeyboardWithListener or something

dpisanu commented:
Have a look at the latest changes.
I think this is where you were going at with your proposed changes.

JakeGinnivan commented:
I'm heading to bed, I might have a small go at this tomorrow to show what I mean. I don't want to waste your time and the 4th revision was good, I just think we can make it slightly better (but unsure until I see the code).

Thanks

dpisanu commented:
Looking forward to your changes.

dpisanu commented:
anything new from your side @JakeGinnivan

dpisanu commented:
final verdict on this change so we can move on to the next tasks?

JakeGinnivan commented:
I will get this in, I just want to have a crack at trying to simplify a bit.

Feel free to start on the next thing, if we merge that one before this one I will rebase and fix up any conflicts.

JakeGinnivan commented:
Just am doing GitVersion stuff today, trying to get the next release out :)

dpisanu commented:
Ok. So'll get cracking on some other things then.

JakeGinnivan commented:
Any thoughts on what next?

JakeGinnivan commented:
Did you see @ivan-danilov's suggestions on how to fix the COM UIA branch in the gitter channel?

dpisanu commented:
I would like to take out Silverlight next. 
Then get Classes and Interfaces in synch.
On one side 
a) to get the code documentation up to date
 and
b) to get a better feel of all the inner workings (something I want for myself before I touch anything COM WRAPPER related with a 10 foot pole)

maxinfet commented:
What ever happened to this pull? It looks like this was ok based on @JakeGinnivan's last few comments.

maxinfet commented:
I ran the tests on this before I went to bed last night and everything looked fine. @JakeGinnivan @Roemer   and @dpisanu give me the heads up if I should not merge this otherwise I am going to merge this on Sunday.
